### PR TITLE
feat(ivy): update specification to include template variables

### PR DESCRIPTION
**Pull request by chuckjaz**
## PR Type
What kind of change does this PR introduce?

```
[x] Feature
```

## What is the current behavior?

No specification how to generate variables.

## What is the new behavior?

Template variables is specified.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 1]
```diff
@@ -193,6 +193,190 @@ describe('compiler specification', () => {
...
+        template: `<ul><li *for="let item of items">{{item.name}}</li></ul>`
+      })
+      class MyComponent {
+        items = [{name: 'one'}, {name: 'two'}];
+
+        // NORMATIVE
+        static ngComponentDef = r3.defineComponent({
+          tag: 'my-component',
+          factory: () => new MyComponent(),
+          template: function(ctx: MyComponent, cm: boolean) {
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: `function MyComponentTemplate`

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 2]
```diff
@@ -193,6 +193,190 @@ describe('compiler specification', () => {
...
+              r3.E(0, 'ul');
+              r3.C(1, c1_dirs, C1);
+              r3.e();
+            }
+            r3.p(1, 'forOf', r3.b(ctx.items));
+            r3.cR(1);
+            ForOfDirective.ngDirectiveDef.r(2, 1);
+            r3.cr();
+
+            function C1(ctx1: any, cm: boolean) {
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: `C1` => `MyComponent_NgForOfDirective_1Template`

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 3]
```diff
@@ -193,6 +193,190 @@ describe('compiler specification', () => {
...
+        items: Item[] = [
+          {name: 'one', infos: [{description: '11'}, {description: '12'}]},
+          {name: 'two', infos: [{description: '21'}, {description: '22'}]}
+        ];
+
+        // NORMATIVE
+        static ngComponentDef = r3.defineComponent({
+          tag: 'my-component',
+          factory: () => new MyComponent(),
+          template: function(ctx: MyComponent, cm: boolean) {
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: `function MyComponentTemplate`

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 4]
```diff
@@ -193,6 +193,190 @@ describe('compiler specification', () => {
...
+              r3.E(0, 'ul');
+              r3.C(1, c1_dirs, C1);
+              r3.e();
+            }
+            r3.p(1, 'forOf', r3.b(ctx.items));
+            r3.cR(1);
+            ForOfDirective.ngDirectiveDef.r(2, 1);
+            r3.cr();
+
+            function C1(ctx1: any, cm: boolean) {
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: `C1` => `MyComponent_NgForOfDirective_1Template`

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 5]
```diff
@@ -193,6 +193,190 @@ describe('compiler specification', () => {
...
+                r3.e();
+              }
+              const l0_item = ctx1.$implicit;
+              r3.t(1, r3.b1('', l0_item.name, ''));
+              r3.p(4, 'forOf', r3.b(ctx.items));
+              r3.cR(3);
+              ForOfDirective.ngDirectiveDef.r(4, 3);
+              r3.cr();
+
+              function C2(ctx2: any, cm: boolean) {
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: `C2` => `MyComponent_NgForOfDirective_NgForOfDirective_2Template`

**Review by mhevery — CHANGES_REQUESTED** [review]

File: packages/core/test/render3/compiler_canonical_spec.ts (hunk 2 above)
Comment (chuckjaz) [reply packages/core/test/render3/compiler_canonical_spec.ts]: How about `MyComponent_NgForOfDirective_Template_1`. The infixed number looks weird to me.

Comment (mhevery) [reply packages/core/test/render3/compiler_canonical_spec.ts]: LGTM

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 6]
```diff
@@ -193,6 +193,190 @@ describe('compiler specification', () => {
...
+              r3.E(0, 'ul');
+              r3.C(1, c1_dirs, MyComponent_ForOfDirective_1);
+              r3.e();
+            }
+            r3.p(1, 'forOf', r3.b(ctx.items));
+            r3.cR(1);
+            ForOfDirective.ngDirectiveDef.r(2, 1);
+            r3.cr();
+
+            function MyComponent_ForOfDirective_1(ctx1: any, cm: boolean) {
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: Why not end with `_Template`?

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 7]
```diff
@@ -193,6 +193,190 @@ describe('compiler specification', () => {
...
+        selector: 'my-component',
+        template: `<ul><li *for="let item of items">{{item.name}}</li></ul>`
+      })
+      class MyComponent {
+        items = [{name: 'one'}, {name: 'two'}];
+
+        // NORMATIVE
+        static ngComponentDef = r3.defineComponent({
+          tag: 'my-component',
+          factory: () => new MyComponent(),
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: For factory method we said we needed to have a name. Do you want to do it now or later PR? (same for refresh and host)

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 8]
```diff
@@ -193,6 +193,190 @@ describe('compiler specification', () => {
...
+            // TODO(chuckj): Not implemented yet
+            // this.view.createEmbeddedView(this.template, context);
+          }
+          this.previous = [...this.forOf];
+        }
+      }
+
+      // NORMATIVE
+      static ngDirectiveDef = r3.defineDirective({
+        factory: () => new ForOfDirective(r3.injectViewContainerRef(), r3.injectTemplateRef()),
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: missing name

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 9]
```diff
@@ -193,6 +193,190 @@ describe('compiler specification', () => {
...
+        // TODO(chuckj): Enable when ngForOf enabling lands.
+        // features: [NgOnChangesFeature(NgForOf)],
+        refresh: (directiveIndex: number, elementIndex: number) => {
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: needs a name

**Review by mhevery — CHANGES_REQUESTED** [review]

**Review by mhevery — APPROVED** [review]

**Pull request closed without merging**